### PR TITLE
[REX] Implement Dino DNA

### PR DESCRIPTION
**Pull request by jimga150**
#11240

### CODE REVIEW

File: Mage.Sets/src/mage/cards/d/DinoDNA.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                null, 6, 6, false
+        );
+        effect.setOnlyColor(ObjectColor.GREEN);
+        effect.setOnlySubType(SubType.DINOSAUR);
+        effect.addAdditionalAbilities(TrampleAbility.getInstance());
+        effect.setText("Create a token that's a copy of target creature card exiled with Dino DNA, except it's a 6/6 green Dinosaur creature with trample.");
+        Ability copyAbility = new ActivateAsSorceryActivatedAbility(Zone.BATTLEFIELD, effect, new ManaCostsImpl<>("{6}"));
+        copyAbility.setTargetAdjuster(DinoDNACopyAdjuster.instance);
+
+        this.addAbility(imprintAbility);
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/d/DinoDNA.java:54]: Must keep ability code in the same block, e.g. add it after prepare above, not at the end.

File: Mage.Sets/src/mage/cards/d/DinoDNA.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        // {6}: Create a token that's a copy of target creature card exiled with Dino DNA, except it's a 6/6 green Dinosaur creature with trample. Activate only as a sorcery.
+        // Based on Croaking Counterpart and Bronzebeak Foragers
+        CreateTokenCopyTargetEffect effect = new CreateTokenCopyTargetEffect(
+                null, null, false, 1, false, false,
+                null, 6, 6, false
+        );
+        effect.setOnlyColor(ObjectColor.GREEN);
+        effect.setOnlySubType(SubType.DINOSAUR);
+        effect.addAdditionalAbilities(TrampleAbility.getInstance());
+        effect.setText("Create a token that's a copy of target creature card exiled with Dino DNA, except it's a 6/6 green Dinosaur creature with trample.");
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/d/DinoDNA.java:52]: {this}

File: Mage.Sets/src/mage/cards/d/DinoDNA.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    public DinoDNA(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.ARTIFACT}, "{1}");
+        
+
+        // Imprint -- {1}, {T}: Exile target creature card from a graveyard. Activate only as a sorcery.
+        // Based on Dimir Doppleganger
+        Ability imprintAbility = new ActivateAsSorceryActivatedAbility(Zone.BATTLEFIELD, new DinoDNAImprintEffect(), new ManaCostsImpl<>("{1}"));
+        imprintAbility.addCost(new TapSourceCost());
+        imprintAbility.addTarget(new TargetCardInGraveyard(new FilterCreatureCard("creature card in a graveyard")));
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/d/DinoDNA.java:39]: StaticFilters.FILTER_CARD_CREATURE_A_GRAVEYARD

File: Mage.Sets/src/mage/cards/d/DinoDNA.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    private DinoDNA(final DinoDNA card) {
+        super(card);
+    }
+
+    @Override
+    public DinoDNA copy() {
+        return new DinoDNA(this);
+    }
+}
+class DinoDNAImprintEffect extends OneShotEffect {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/d/DinoDNA.java:68]: (style: empty line between closing brace and new class)

File: Mage.Sets/src/mage/cards/d/DinoDNA.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    DinoDNAImprintEffect() {
+        super(Outcome.Benefit);
+        staticText = "Exile target creature card from a graveyard.";
+    }
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/d/DinoDNA.java:73]: hm, does `new ExileTargetEffect().setToSourceExileZone(true)` not accomplish what you need?

Comment (jimga150) [reply Mage.Sets/src/mage/cards/d/DinoDNA.java:73]: ah, nice! the examples i was going off of werent using that, that totally does it lol

**Comment by jimga150** [comment]
@xenohedron rereading the effect, it looks like it also causes creature cards that have additional supertypes to lose those supertypes. Thats how i would read it, but i cant find any rulings stating as much. it looks like based on cards like [Myrkul, Lord of Bones](https://scryfall.com/card/clb/287/myrkul-lord-of-bones), this would need a whole new effect for this.

**Comment by jimga150** [comment]
> @xenohedron rereading the effect, it looks like it also causes creature cards that have additional supertypes to lose those supertypes. Thats how i would read it, but i cant find any rulings stating as much. it looks like based on cards like [Myrkul, Lord of Bones](https://scryfall.com/card/clb/287/myrkul-lord-of-bones), this would need a whole new effect for this.

scratch that, this uses an inline effect modifier i could use instead of making a new class. question stands

**Comment by xenohedron** [comment]
205.4b. An [object](https://yawgatog.com/resources/magic-rules/#object)'s [supertype](https://yawgatog.com/resources/magic-rules/#supertype) is [independent](https://yawgatog.com/resources/magic-rules/#independent) of its [card type](https://yawgatog.com/resources/magic-rules/#card_type) and [subtype](https://yawgatog.com/resources/magic-rules/#subtype), even though some [supertypes](https://yawgatog.com/resources/magic-rules/#supertype) are closely identified with specific [card types](https://yawgatog.com/resources/magic-rules/#card_type). Changing an [object](https://yawgatog.com/resources/magic-rules/#object)'s [card types](https://yawgatog.com/resources/magic-rules/#card_type) or [subtypes](https://yawgatog.com/resources/magic-rules/#subtype) won't change its [supertypes](https://yawgatog.com/resources/magic-rules/#supertype). Changing an [object](https://yawgatog.com/resources/magic-rules/#object)'s [supertypes](https://yawgatog.com/resources/magic-rules/#supertype) won't change its [card types](https://yawgatog.com/resources/magic-rules/#card_type) or [subtypes](https://yawgatog.com/resources/magic-rules/#subtype). When an [object](https://yawgatog.com/resources/magic-rules/#object) gains or loses a [supertype](https://yawgatog.com/resources/magic-rules/#supertype), it retains any other [supertypes](https://yawgatog.com/resources/magic-rules/#supertype) it had.

**Review by xenohedron — APPROVED** [review]

**Comment by jimga150** [comment]
> 205.4b. An [object](https://yawgatog.com/resources/magic-rules/#object)'s [supertype](https://yawgatog.com/resources/magic-rules/#supertype) is [independent](https://yawgatog.com/resources/magic-rules/#independent) of its [card type](https://yawgatog.com/resources/magic-rules/#card_type) and [subtype](https://yawgatog.com/resources/magic-rules/#subtype), even though some [supertypes](https://yawgatog.com/resources/magic-rules/#supertype) are closely identified with specific [card types](https://yawgatog.com/resources/magic-rules/#card_type). Changing an [object](https://yawgatog.com/resources/magic-rules/#object)'s [card types](https://yawgatog.com/resources/magic-rules/#card_type) or [subtypes](https://yawgatog.com/resources/magic-rules/#subtype) won't change its [supertypes](https://yawgatog.com/resources/magic-rules/#supertype). Changing an [object](https://yawgatog.com/resources/magic-rules/#object)'s [supertypes](https://yawgatog.com/resources/magic-rules/#supertype) won't change its [card types](https://yawgatog.com/resources/magic-rules/#card_type) or [subtypes](https://yawgatog.com/resources/magic-rules/#subtype). When an [object](https://yawgatog.com/resources/magic-rules/#object) gains or loses a [supertype](https://yawgatog.com/resources/magic-rules/#supertype), it retains any other [supertypes](https://yawgatog.com/resources/magic-rules/#supertype) it had.

very interesting

**Pull request merged**